README.md
[Github](https://github.com/LedgerHQ/ledgerjs/),
[Ledger Devs Slack](https://ledger-dev.slack.com/)

## @obsidiansystems/hw-app-ckb

Ledger Hardware Wallet CKB JavaScript bindings.

## Maintainers and packagers

Please see the [changelog](./ChangeLog.md) and [release docs](./Releasing.md)
for this package.

## Using LedgerJS for CKB

Here is a sample app for Node:

    let Transport = require("@ledgerhq/hw-transport-node-hid").default;
    let Ckb = require("@obsidiansystems/hw-app-ckb").default;

    const getWalletId = async () => {
      const transport = await Transport.create();
      const ckb = new Ckb(transport);
      const result = await ckb.getWalletId();
      return result;
    };

    const signTransaction = async () => {
      const transport = await Transport.create();
      const ckb = new Ckb(transport);
      contexts = [ "010000000d0200001c000000200000006e00000092000000ee000000f10100000000000002000000a563884b3686078ec7e7677a5f86449b15cf2693f3c1241766c6996f206cc5410200000000ace5ea83c478bb866edf122ff862085789158f5cbff155b7bb5f13058555b708000000000101000000327f1fc62c53530c6c27018f1e8cee27c35c0370c3b4d3376daf8fe110e7d8cb020000000000000000000000c399495011b912999dbc72cf54982924e328ae170654ef76c8aba190ca376307000000000000000000000000c317d0b0b2a513ab1206e6d454c1960de7d7b4b80d0748a3e1f9cb197b74b8a501000000030100000c000000a20000009600000010000000180000006100000000e8764817000000490000001000000030000000310000009bd7e06f3ecf4be0f2fcd2188b23f1b9fcc88e5d4b65a8637b17723bbda3cce80114000000e5260d839a786ac2a909181df9a423f1efbe863d3500000010000000300000003100000082d76d1b75fe2fd9a27dfbaa65a039221a380d76c926f378d3f81cf3e7e13f2e01000000006100000010000000180000006100000064e5b27317000000490000001000000030000000310000009bd7e06f3ecf4be0f2fcd2188b23f1b9fcc88e5d4b65a8637b17723bbda3cce80114000000e5260d839a786ac2a909181df9a423f1efbe863d1c0000000c0000001800000008000000520700000000000000000000"
                 , "00000000570100001c000000200000006e000000b2000000e20000004b0100000000000002000000a563884b3686078ec7e7677a5f86449b15cf2693f3c1241766c6996f206cc5410200000000ace5ea83c478bb866edf122ff862085789158f5cbff155b7bb5f13058555b708000000000102000000327f1fc62c53530c6c27018f1e8cee27c35c0370c3b4d3376daf8fe110e7d8cb4930ba433e606a53f4f283f02dddeb6d51b0dc3e463629b14a27995de9c71eca01000000ba08000000010020b1b547956a0dfb7ea618231563b3acd23607586e939f88e5a6db5f392b2e78d500000000690000000800000061000000100000001800000061000000c561436317000000490000001000000030000000310000009bd7e06f3ecf4be0f2fcd2188b23f1b9fcc88e5d4b65a8637b17723bbda3cce80114000000e5260d839a786ac2a909181df9a423f1efbe863d0c0000000800000000000000" ];
      transaction = "d00100000c000000bc010000b00100001c000000200000006e00000072000000ce0000009c0100000000000002000000a563884b3686078ec7e7677a5f86449b15cf2693f3c1241766c6996f206cc5410200000000ace5ea83c478bb866edf122ff862085789158f5cbff155b7bb5f13058555b708000000000100000000020000000000000000000000b1b547956a0dfb7ea618231563b3acd23607586e939f88e5a6db5f392b2e78d5010000000000000000000000258e82bab2af21fd8899fc872742f4acea831f5e4c232297816b9bf4a19597a900000000ce0000000c0000006d0000006100000010000000180000006100000000e8764817000000490000001000000030000000310000009bd7e06f3ecf4be0f2fcd2188b23f1b9fcc88e5d4b65a8637b17723bbda3cce80114000000e5260d839a786ac2a909181df9a423f1efbe863d61000000100000001800000061000000e91c708e17000000490000001000000030000000310000009bd7e06f3ecf4be0f2fcd2188b23f1b9fcc88e5d4b65a8637b17723bbda3cce80114000000e5260d839a786ac2a909181df9a423f1efbe863d140000000c000000100000000000000000000000140000000c000000100000000000000000000000"
      const result = await ckb.signTransaction("44'/309'/0'/1/0", transaction, null, contexts, "44'/309'/0'/1/0");
      return result;
    };

    const getVersion = async () => {
      const transport = await Transport.create();
      const ckb = new Ckb(transport);
      const versionInfo = await ckb.getAppConfiguration();
      return versionInfo;
    };

    const getAddress = async () => {
      const transport = await Transport.create();
      const ckb = new Ckb(transport);
      const result = await ckb.getWalletPublicKey("44'/309'/0'/1/0");
      return result;
    };

    const doAll = async () => {
        walletId = await getWalletId();
        console.log(walletId);
        version = await getVersion();
        console.log(version);
        address = await getAddress();;
        console.log(address);
        signature = await signTransaction();
        console.log(signature);
    };

    doAll().catch(err => console.log(err));

## API

<!-- Generated by documentation.js. Update this documentation by updating the source code. -->

#### Table of Contents

-   [Ckb](#ckb)
    -   [Parameters](#parameters)
    -   [Examples](#examples)
    -   [getWalletPublicKey](#getwalletpublickey)
        -   [Parameters](#parameters-1)
        -   [Examples](#examples-1)
    -   [getWalletExtendedPublicKey](#getwalletextendedpublickey)
        -   [Parameters](#parameters-2)
        -   [Examples](#examples-2)
    -   [signTransaction](#signtransaction)
        -   [Parameters](#parameters-3)
        -   [Examples](#examples-3)
    -   [buildAnnotatedTransaction](#buildannotatedtransaction)
        -   [Parameters](#parameters-4)
    -   [signAnnotatedTransaction](#signannotatedtransaction)
        -   [Parameters](#parameters-5)
    -   [defaultSighashWitness](#defaultsighashwitness)
    -   [getAppConfiguration](#getappconfiguration)
        -   [Examples](#examples-4)
    -   [getWalletId](#getwalletid)
        -   [Examples](#examples-5)

### Ckb

Nervos API

#### Parameters

-   `transport` **Transport&lt;any>** 
-   `scrambleKey` **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)**  (optional, default `"CKB"`)

#### Examples

```javascript
import Ckb from "@obsidiansystems/hw-app-ckb";
const ckb = new Ckb(transport);
```

#### getWalletPublicKey

get CKB address for a given BIP 32 path.

##### Parameters

-   `path` **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** a path in BIP 32 format
-   `testnet` **[boolean](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Boolean)** 

##### Examples

```javascript
const result = await ckb.getWalletPublicKey("44'/144'/0'/0/0");
const publicKey = result.publicKey;
const lockArg = result.lockArg;
const address = result.address;
```

Returns **[Promise](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Promise)&lt;[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)>** an object with a publicKey, lockArg, and (secp256k1+blake160) address.

#### getWalletExtendedPublicKey

get extended public key for a given BIP 32 path.

##### Parameters

-   `path` **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** a path in BIP 32 format

##### Examples

```javascript
const result = await ckb.getWalletPublicKey("44'/144'/0'/0/0");
const publicKey = result;
```

Returns **[Promise](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Promise)&lt;[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)>** an object with a publicKey

#### signTransaction

Sign a Nervos transaction with a given BIP 32 path

##### Parameters

-   `signPath` **([string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String) | BIPPath | \[[number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number)])** the path to sign with, in BIP 32 format
-   `rawTx` **([string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String) | blockchain.RawTransactionJSON)** 
-   `groupWitnessesHex` **\[[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)]?** hex of in-group and extra witnesses to include in signature
-   `rawContextsTx` **\[([string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String) | blockchain.RawTransactionJSON)]** 
-   `changePath` **([string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String) | BIPPath | \[[number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number)])** the path the transaction sends change to, in BIP 32 format (optional, defaults to signPath)
-   `rawTxHex`  transaction to sign
-   `contextTransaction`  list of transaction contexts for parsing

##### Examples

```javascript
TODO
```

Returns **[Promise](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Promise)&lt;[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)>** a signature as hex string

#### buildAnnotatedTransaction

Construct an AnnotatedTransaction for a given collection of signing data

Parameters are the same as for signTransaction, but no ledger interaction is attempted.

AnnotatedTransaction is a type defined for the ledger app that collects
all of the information needed to securely confirm a transaction on-screen
and a few bits of duplicative information to allow it to be processed as a
stream.

##### Parameters

-   `signPath` **([string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String) | BIPPath | \[[number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number)])** 
-   `rawTx` **([string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String) | RawTransactionJSON)** 
-   `groupWitnesses` **\[[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)]?** 
-   `rawContextsTx` **\[([string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String) | RawTransactionJSON)]** 
-   `changePath` **([string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String) | BIPPath | \[[number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number)])** 

Returns **AnnotatedTransactionJSON** 

#### signAnnotatedTransaction

Sign an already constructed AnnotatedTransaction.

##### Parameters

-   `tx` **(AnnotatedTransaction | AnnotatedTransactionJSON)** 

Returns **[Promise](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Promise)&lt;[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)>** 

#### defaultSighashWitness

An empty WitnessArgs with enough space to fit a sighash signature into.

#### getAppConfiguration

Get the version of the Nervos app installed on the hardware device

##### Examples

```javascript
const result = await ckb.getAppConfiguration();

{
  "version": "1.0.3",
  "hash": "0000000000000000000000000000000000000000"
}
```

Returns **[Promise](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Promise)&lt;{version: [string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String), hash: [string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)}>** an object with a version

#### getWalletId

Get the wallet identifier for the Ledger wallet

##### Examples

```javascript
const id = await ckb.getWalletId();

"0x69c46b6dd072a2693378ef4f5f35dcd82f826dc1fdcc891255db5870f54b06e6"
```

Returns **[Promise](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Promise)&lt;[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)>** a byte string


ChangeLog.md
# Revision history for `hw-app-ckb`

This project's release branch is `master`.
This log is written from the perspective of the release branch: when changes hit `master`, they are considered released.

## v0.1.2 - 2020-11-18

Includes the bug fix merged in [!10](https://github.com/obsidiansystems/hw-app-ckb/pull/10)

This release has been tested with:
- Nervos Ledger App: [v0.5.1 - 61435fe19c722a0445ac3743cc42b84139842c33](https://github.com/obsidiansystems/ledger-app-nervos/releases/tag/v0.5.1)

## v0.1.1 - 2020-10-26

Primarily a bugfix release.

- All compiled files are now included in release tarballs after `yarn pack`.
- A specific version of Nixpkgs is now pinned in `shell.nix` for developer
  convenience.
- Ledger app version hash parsing in `getAppConfiguration` has been fixed.
- Release process documentation is now included.

## v0.1.0 - 2020-09-30

This is the first release since moving this to its own repo from https://github.com/obsidiansystems/ledgerjs/tree/nervos.

### Release Notes
- Small discrepancies in the Nervos schemas between the RPC and Lumos have been worked around.
- Added `package.yaml`s for examples project and for example workspace.

This release has been tested with:
- Nervos Ledger App: [v0.5.0 - 88026362a0bbf096ae911f33be5149415a2a7c77](https://github.com/obsidiansystems/ledger-app-nervos/releases/tag/v0.5.0)
- CKB-CLI: [b460c998d6681a89a47b3af203ecc5f12d7b2507](https://github.com/obsidiansystems/ckb-cli/commit/b460c998d6681a89a47b3af203ecc5f12d7b2507)


examples/ckb-sdk-js-example.js
/* eslint-disable */

/* NOTE: This file needs a package.json to provide the dependencies */

const util = require('util')
const CKB = require('@obsidiansystems/hw-app-ckb').default
const formatter = require('@nervosnetwork/ckb-sdk-core/paramsFormatter').default

// Whether to connect to a running instance of the Speculos simulator for
// ledger apps or a real physical ledger
const useSpeculos = false

// Note: there is currently a bug in the speculos simulator that appears not to
// have been adequately addressed that prevents correctly generating the the
// last byte of a recoverable signature. You can work around this by replacing
// the last byte with 00, trying to send the signature, and if that fails
// trying again with 01. This issue _ONLY_ occurs with Speculos, which has a
// different implementation of cryptography primitives than the real hardware.

let Transport = null
if ( useSpeculos ) {
// For speculos:
  Transport = require("@ledgerhq/hw-transport-node-speculos").default;
} else {
// For a real ledger:
  Transport = require("@ledgerhq/hw-transport-node-hid").default;
}
console.log(Transport)

const LedgerCkb = require("@obsidiansystems/hw-app-ckb").default;
const blk = require("@obsidiansystems/hw-app-ckb/annotated.js");

const bootstrap = async () => {
  const nodeUrl = process.env.NODE_URL || 'http://localhost:8114' // example node url
  const blockAssemblerCodeHash = '0x9bd7e06f3ecf4be0f2fcd2188b23f1b9fcc88e5d4b65a8637b17723bbda3cce8' // transcribe the block_assembler.code_hash in the ckb.toml from the ckb project

  const ckb = new CKB(nodeUrl) // instantiate the JS SDK with provided node url

  await ckb.loadDeps() // load the dependencies of secp256k1 algorithm which is used to verify the signature in transaction's witnesses.

  let transport = null
  if ( useSpeculos ) {
    // To connect to a speculos instance:
    apduPort = 9999;
    transport = await Transport.open( { apduPort } );
  } else {
    transport = await Transport.open();
  }

  const lckb = new LedgerCkb(transport)
  
  const keydata = await lckb.getWalletPublicKey("44'/309'/0'/0/0", true)
  const publicKeyHash = "0x" + keydata.lockArg
  const address = keydata.address
  addresses = { testnetAddress: address }
  
  /**
   * to see the addresses
   */
  // console.log(JSON.stringify(addresses, null, 2))

  /**
   * calculate the lockHash by the address publicKeyHash
   * 1. the publicKeyHash of the address is required in the args field of lock script
   * 2. compose the lock script with the code hash(as a miner, we use blockAssemblerCodeHash here), and args
   * 3. calculate the hash of lock script via ckb.utils.scriptToHash method
   */

  const lockScript = {
    hashType: "type",
    codeHash: blockAssemblerCodeHash,
    args: publicKeyHash,
  }

  /**
   * to see the lock script
   */
  // console.log(JSON.stringify(lockScript, null, 2))

  const lockHash = ckb.utils.scriptToHash(lockScript)

  /**
   * to see the lock hash
   */
  // console.log(lockHash)

  // method to fetch all unspent cells by lock hash
  const unspentCells = await ckb.loadCells({
    lockHash
  })

  /**
   * to see the unspent cells
   */
  // console.log(unspentCells)

  /**
   * send transaction
   */
  const toAddress = ckb.utils.privateKeyToAddress("0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff", {
    prefix: 'ckt'
  })

  const rawTransaction = ckb.generateRawTransaction({
    fromAddress: addresses.testnetAddress,
    toAddress,
    capacity: BigInt(600000000000),
    fee: BigInt(100000),
    safeMode: true,
    cells: unspentCells,
    deps: ckb.config.secp256k1Dep,
  })

  rawTransaction.witnesses = rawTransaction.inputs.map(() => '0x')
  rawTransaction.witnesses[0] = ckb.utils.serializeWitnessArgs({
    lock: '',
    inputType: '',
    outputType: ''
  })

  // fetch all the context transactions
  ctxds = (await Promise.all(rawTransaction.inputs.map(a=>ckb.rpc.getTransaction(a.previousOutput.txHash)))).map(a=>a.transaction)

  const formatted = formatter.toRawTransaction(rawTransaction)
  const formattedCtxd = ctxds.map(formatter.toRawTransaction)

  signature = await lckb.signTransaction("44'/309'/0'/0/0", formatted, formatted.witnesses, formattedCtxd, "44'/309'/0'/0/0")
  rawTransaction.witnesses[0] = ckb.utils.serializeWitnessArgs( { lock: "0x"+signature, inputType: '', outputType: '' });

  const realTxHash = await ckb.rpc.sendTransaction(rawTransaction).catch(err=>err)

  /**
   * to see the real transaction hash
   */
  console.log(`The real transaction hash is: ${realTxHash}`)
}

bootstrap()


src/Ckb.js
//@flow

import type Transport from "@ledgerhq/hw-transport";
import BIPPath from "bip32-path";
import * as blockchain from "./annotated";
import Blake2b from "blake2b-wasm";
import * as bech32 from "bech32";

/**
 * Nervos API
 *
 * @example
 * import Ckb from "@obsidiansystems/hw-app-ckb";
 * const ckb = new Ckb(transport);
 */
export default class Ckb {
  transport: Transport<*>;

  constructor(transport: Transport<*>, scrambleKey: string = "CKB") {
    this.transport = transport;
    transport.decorateAppAPIMethods(
      this,
      [
        "getAppConfiguration",
        "getWalletId",
        "getWalletPublicKey",
        "signAnnotatedTransaction",
      ],
      scrambleKey
    );
  }

  /**
   * get CKB address for a given BIP 32 path.
   *
   * @param path a path in BIP 32 format
   * @return an object with a publicKey, lockArg, and (secp256k1+blake160) address.
   * @example
   * const result = await ckb.getWalletPublicKey("44'/144'/0'/0/0");
   * const publicKey = result.publicKey;
   * const lockArg = result.lockArg;
   * const address = result.address;
   */
  async getWalletPublicKey(path: string, testnet: boolean): Promise<string> {
    const bipPath = BIPPath.fromString(path).toPathArray();

    const cla = 0x80;
    const ins = 0x02;
    const p1 = 0x00;
    const p2 = 0x00;
    const data = Buffer.alloc(1 + bipPath.length * 4);

    data.writeUInt8(bipPath.length, 0);
    bipPath.forEach((segment, index) => {
      data.writeUInt32BE(segment, 1 + index * 4);
    });

    const response = await this.transport.send(cla, ins, p1, p2, data);

    const publicKeyLength = response[0];
    const publicKey = response.slice(1, 1 + publicKeyLength);

    const compressedPublicKey = Buffer.alloc(33);
    compressedPublicKey.fill(publicKey[64] & 1 ? "03" : "02", 0, 1, "hex");
    compressedPublicKey.fill(publicKey.subarray(1, 33), 1, 33);
    const hashPersonalization = Uint8Array.from([99, 107, 98, 45, 100, 101, 102, 97, 117, 108, 116, 45, 104, 97, 115, 104]);
    const lockArg = Buffer.from(
      Blake2b(32, null, null, hashPersonalization)
        .update(compressedPublicKey)
        .digest("binary")
        .subarray(0, 20)
    );

    const addr_contents = Buffer.alloc(22);
    addr_contents.fill("0100", 0, 2, "hex");
    addr_contents.fill(lockArg, 2, 22);
    const addr = bech32.encode(
      testnet ? "ckt" : "ckb",
      bech32.toWords(addr_contents)
    );

    return {
      publicKey: publicKey.toString("hex"),
      lockArg: lockArg.toString("hex"),
      address: addr,
    };
  }

  /**
   * get extended public key for a given BIP 32 path.
   *
   * @param path a path in BIP 32 format
   * @return an object with a publicKey
   * @example
   * const result = await ckb.getWalletPublicKey("44'/144'/0'/0/0");
   * const publicKey = result;
   */
  async getWalletExtendedPublicKey(path: string): Promise<string> {
    const bipPath = BIPPath.fromString(path).toPathArray();

    const cla = 0x80;
    const ins = 0x04;
    const p1 = 0x00;
    const p2 = 0x00;
    const data = Buffer.alloc(1 + bipPath.length * 4);

    data.writeUInt8(bipPath.length, 0);
    bipPath.forEach((segment, index) => {
      data.writeUInt32BE(segment, 1 + index * 4);
    });

    const response = await this.transport.send(cla, ins, p1, p2, data);
    const publicKeyLength = response[0];
    const chainCodeOffset = 2 + publicKeyLength;
    const chainCodeLength = response[1 + publicKeyLength];
    return {
      public_key: response.slice(1, 1 + publicKeyLength).toString("hex"),
      chain_code: response
        .slice(chainCodeOffset, chainCodeOffset + chainCodeLength)
        .toString("hex"),
    };
  }

  /**
   * Sign a Nervos transaction with a given BIP 32 path
   *
   * @param signPath the path to sign with, in BIP 32 format
   * @param rawTxHex transaction to sign
   * @param groupWitnessesHex hex of in-group and extra witnesses to include in signature
   * @param contextTransaction list of transaction contexts for parsing
   * @param changePath the path the transaction sends change to, in BIP 32 format (optional, defaults to signPath)
   * @return a signature as hex string
   * @example
   * TODO
   */

  async signTransaction(
    signPath: string | BIPPath | [number],
    rawTx: string | blockchain.RawTransactionJSON,
    groupWitnessesHex?: [string],
    rawContextsTx: [string | blockchain.RawTransactionJSON],
    changePath: string | BIPPath | [number]
  ): Promise<string> {
    return await this.signAnnotatedTransaction(
      this.buildAnnotatedTransaction(
        signPath,
        rawTx,
        groupWitnessesHex,
        rawContextsTx,
        changePath
      )
    );
  }

  /**
   * Construct an AnnotatedTransaction for a given collection of signing data
   *
   * Parameters are the same as for signTransaction, but no ledger interaction is attempted.
   *
   * AnnotatedTransaction is a type defined for the ledger app that collects
   * all of the information needed to securely confirm a transaction on-screen
   * and a few bits of duplicative information to allow it to be processed as a
   * stream.
   */

  buildAnnotatedTransaction(
    signPath: string | BIPPath | [number],
    rawTx: string | RawTransactionJSON,
    groupWitnesses?: [string],
    rawContextsTx: [string | RawTransactionJSON],
    changePath: string | BIPPath | [number]
  ): AnnotatedTransactionJSON {
    const prepBipPath = (pathSrc) => {
      if (Array.isArray(pathSrc)) {
        return pathSrc;
      }
      if (typeof pathSrc === "object") {
        return pathSrc.toPathArray();
      }
      if (typeof pathSrc === "string") {
        return BIPPath.fromString(pathSrc).toPathArray();
      }
    };

    const signBipPath = prepBipPath(signPath);
    const changeBipPath = prepBipPath(changePath);

    const getRawTransactionJSON = (rawTrans) => {
      if (typeof rawTrans === "string") {
        const rawTxBuffer = Buffer.from(rawTrans, "hex");
        return new blockchain.RawTransaction(rawTxBuffer.buffer).toObject();
      }
      return rawTrans;
    };

    const contextTransactions = rawContextsTx.map(getRawTransactionJSON);

    const rawTxUnpacked = getRawTransactionJSON(rawTx);

    const annotatedCellInputVec = rawTxUnpacked.inputs.map((inpt, idx) => ({
      input: inpt,
      source: contextTransactions[idx],
    }));

    const annotatedRawTransaction = {
      version: rawTxUnpacked.version,
      cell_deps: rawTxUnpacked.cell_deps,
      header_deps: rawTxUnpacked.header_deps,
      inputs: annotatedCellInputVec,
      outputs: rawTxUnpacked.outputs,
      outputs_data: rawTxUnpacked.outputs_data,
    };

    return {
      signPath: signBipPath,
      changePath: changeBipPath,
      inputCount: rawTxUnpacked.inputs.length,
      raw: annotatedRawTransaction,
      witnesses:
        Array.isArray(groupWitnesses) && groupWitnesses.length > 0
          ? groupWitnesses
          : [this.defaultSighashWitness],
    };
  }

  /**
   * Sign an already constructed AnnotatedTransaction.
   */
  async signAnnotatedTransaction(
    tx: AnnotatedTransaction | AnnotatedTransactionJSON
  ): Promise<string> {
    const rawAnTx = Buffer.from(blockchain.SerializeAnnotatedTransaction(tx));

    const maxApduSize = 230;

    let txFullChunks = Math.floor(rawAnTx.byteLength / maxApduSize);
    let isContinuation = 0x00;
    for (let i = 0; i < txFullChunks; i++) {
      let data = rawAnTx.slice(i * maxApduSize, (i + 1) * maxApduSize);
      await this.transport.send(0x80, 0x03, isContinuation, 0x00, data);
      isContinuation = 0x01;
    }

    let lastOffset = txFullChunks * maxApduSize;
    let lastData = rawAnTx.slice(lastOffset, lastOffset + maxApduSize);
    let response = await this.transport.send(
      0x80,
      0x03,
      isContinuation | 0x80,
      0x00,
      lastData
    );
    return response.slice(0,65).toString("hex");
  }

  /**
   * An empty WitnessArgs with enough space to fit a sighash signature into.
   */
  defaultSighashWitness =
    "55000000100000005500000055000000410000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000";

  /**
   * Get the version of the Nervos app installed on the hardware device
   *
   * @return an object with a version
   * @example
   * const result = await ckb.getAppConfiguration();
   *
   * {
   *   "version": "1.0.3",
   *   "hash": "0000000000000000000000000000000000000000"
   * }
   */
  async getAppConfiguration(): Promise<{
    version: string,
    hash: string,
  }> {
    const response1 = await this.transport.send(0x80, 0x00, 0x00, 0x00);
    const response2 = await this.transport.send(0x80, 0x09, 0x00, 0x00);

    const result = {};
    result.version =
      "" + response1[0] + "." + response1[1] + "." + response1[2];
    result.hash = response2.slice(0,-3).toString(); // Strip off null and 0x9000

    return result;
  }

  /**
   * Get the wallet identifier for the Ledger wallet
   *
   * @return a byte string
   * @example
   * const id = await ckb.getWalletId();
   *
   * "0x69c46b6dd072a2693378ef4f5f35dcd82f826dc1fdcc891255db5870f54b06e6"
   */
  async getWalletId(): Promise<string> {
    const response = await this.transport.send(0x80, 0x01, 0x00, 0x00);

    const result = response.slice(0, 32).toString("hex");

    return result;
  }

  async signMessage(
    path: string,
    rawMsgHex: string,
    displayHex: bool
  ): Promise<string> {
    const bipPath = BIPPath.fromString(path).toPathArray();
    const magicBytes = Buffer.from("Nervos Message:");
    const rawMsg = Buffer.concat([magicBytes, Buffer.from(rawMsgHex, "hex")]);

    //Init apdu
    let rawPath = Buffer.alloc(1 + 1 + bipPath.length * 4);
    rawPath.writeInt8(displayHex, 0);
    rawPath.writeInt8(bipPath.length, 1);
    bipPath.forEach((segment, index) => {
      rawPath.writeUInt32BE(segment, 2 + index * 4);
    });
    await this.transport.send(0x80, 0x06, 0x00, 0x00, rawPath);

    // Msg Chunking
    const maxApduSize = 230;
    let txFullChunks = Math.floor(rawMsg.length / maxApduSize);
    for (let i = 0; i < txFullChunks; i++) {
      let data = rawMsg.slice(i*maxApduSize, (i+1)*maxApduSize);
      await this.transport.send(0x80, 0x06, 0x01, 0x00, data);
    }

    let lastOffset = Math.floor(rawMsg.length / maxApduSize) * maxApduSize;
    let lastData = rawMsg.slice(lastOffset, lastOffset+maxApduSize);
    let response = await this.transport.send(0x80, 0x06, 0x81, 0x00, lastData);
    return response.slice(0,65).toString("hex");
  }

}

